File: cpp/src/leetcode/LC0068TextJustification.cpp
#include "leetcode/LC0068TextJustification.hpp"

#include <numeric>

namespace {
    std::string leftAlign(const std::vector<std::string> &words, int maxWidth);
    std::string justify(const std::vector<std::string> &words, int maxWidth);
}

std::vector<std::string> LC0068TextJustification::fullJustify(const std::vector<std::string> &words, const int maxWidth) {
    std::vector<std::string> result;
    std::vector<std::string> wordsInLine;

    size_t wordsInLineLength = 0;

    for (auto &word : words) {
        if (wordsInLineLength + wordsInLine.size() + word.size() > maxWidth) {
            result.push_back(justify(wordsInLine, maxWidth));

            wordsInLine = {};
            wordsInLineLength = 0;
        }

        wordsInLine.push_back(word);
        wordsInLineLength += word.size();
    }

    if (!wordsInLine.empty()) {
        result.push_back(leftAlign(wordsInLine, maxWidth));
    }

    return result;
}

namespace {
    std::string leftAlign(const std::vector<std::string> &words, const int maxWidth) {
        std::string line = words[0];

        for (int i = 1; i < words.size(); ++i) {
            line += " " + words[i];
        }
        line += std::string(maxWidth - line.size(), ' ');

        return line;
    }

    std::string justify(const std::vector<std::string> &words, const int maxWidth) {
        const size_t wordCount = words.size();
        if (wordCount == 1) {
            return leftAlign(words, maxWidth);
        }

        const size_t charCount = std::accumulate(words.begin(), words.end(), 0,
            [](const size_t count, const std::string &word) { return count + word.size(); });
        const size_t totalSpaceCount = maxWidth - charCount;
        const size_t gapCount = wordCount - 1;
        const size_t baseSpaceCount = totalSpaceCount / gapCount;
        const size_t extraSpaceCount = totalSpaceCount % gapCount;

        std::string line = words[0];

        for (int i = 1; i < wordCount; ++i) {
            const size_t spaceCount = baseSpaceCount + (i <= extraSpaceCount ? 1 : 0);
            line += std::string(spaceCount, ' ');
            line += words[i];
        }

        return line;
    }
}

File: rust/src/leetcode/lc0068_text_justification.rs
#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn testcase_1() {
        let words = ["This", "is", "an", "example", "of", "text", "justification."];
        let max_width = 16;
        let output : Vec<String> = [
            "This    is    an",
            "example  of text",
            "justification.  "
        ].iter().map(|s| s.to_string()).collect();
        assert_eq!(output, full_justify(&words, max_width));
    }

    #[test]
    fn testcase_2() {
        let words = ["What","must","be","acknowledgment","shall","be"];
        let max_width = 16;
        let output : Vec<String> = [
            "What   must   be",
            "acknowledgment  ",
            "shall be        "
        ].iter().map(|s| s.to_string()).collect();
        assert_eq!(output, full_justify(&words, max_width));
    }

    #[test]
    fn testcase_3() {
        let words = ["Science","is","what","we","understand","well","enough","to","explain","to","a",
            "computer.","Art","is","everything","else","we","do"];
        let max_width = 20;
        let output : Vec<String> = [
            "Science  is  what we",
            "understand      well",
            "enough to explain to",
            "a  computer.  Art is",
            "everything  else  we",
            "do                  "
        ].iter().map(|s| s.to_string()).collect();
        assert_eq!(output, full_justify(&words, max_width));
    }
}

pub fn full_justify(words: &[&str], max_width: i32) -> Vec<String> {
    let max_width = max_width as usize;
    let mut result = vec![];
    let mut words_in_line = vec![];
    let mut words_width = 0;

    for word in words {
        let new_width = words_width + word.len() + words_in_line.len();

        if new_width > max_width {
            result.push(fully_justify_line(&words_in_line, max_width));

            words_in_line.clear();
            words_width = 0;
        }

        words_width += word.len();
        words_in_line.push(word);
    }

    if words_in_line.len() > 0 {
        result.push(left_justify_line(&words_in_line, max_width));
    }

    result
}

fn fully_justify_line(words_in_line: &[&str], max_width: usize) -> String {
    let n = words_in_line.len();
    if n == 1 {
        return left_justify_line(words_in_line, max_width);
    }

    let char_count: usize = words_in_line.iter().map(|s| s.len()).sum();
    let total_space_count = max_width - char_count;
    let gap_count = n - 1;
    let base_space_count = total_space_count / gap_count;
    let extra_space_count = total_space_count % gap_count;

    let mut result = String::with_capacity(max_width);

    for (i, word) in words_in_line.iter().enumerate() {
        result.push_str(word);

        if i == gap_count {
            continue;
        }

        let space_count = base_space_count + if i < extra_space_count { 1 } else { 0 };
        for _ in 0..space_count {
            result.push(' ');
        }
    }

    result
}

fn left_justify_line(words_in_line: &[&str], max_width: usize) -> String {
    let mut result = String::with_capacity(max_width);

    for i in 0..words_in_line.len() {
        if i > 0 {
            result.push(' ');
        }
        result.push_str(words_in_line[i]);
    }

    while result.len() < max_width {
        result.push(' ');
    }

    result
}
